### ISSUE TITLE
Float>>printOn:decimalPlaces: is slow

### ISSUE THREAD

**Issue by fxgallego** [dolphinsmalltalk/Dolphin#1026]
**Describe the bug**
I was printing a large table in html, which had a lot of Floats, and noticed that printing floats with decimal places where really slow.
So I started to do some investigation and experiments and found that this expression:

```smalltalk
Time millisecondsToRun: [
    100000 timesRepeat: [
        stream := String writeStream.
        Float pi printOn: stream decimalPlaces: 2.
        stream grabContents
    ]
]
```

returns the following times in milliseconds:

Dolphin original -> 1190 .. 1220
Dolphin sprintf -> 300 .. 370
Pharo port to Dolphin -> 290 .. 330

Additionally, and only to use as metric, in my Chrome console, in JavaScript, runs in 40ms

For the sprintf implementation I added a new sprintf function call variant in VMLibrary supporting passing Float as parameter. I guess that's wrong, and I don't need to do that, but it worked.

The Pharo port was really easy and only required 3 methods:

```smalltalk
!Fraction methodsFor!

printOn: aStream showingDecimalPlaces: placesDesired
	"Same as super, but provides a faster implementation by inlining some Fraction protocol thus avoiding intermediate Fraction creation."
	
	| roundedFractionPart integerPart scaling |
	placesDesired <= 0
		ifTrue: [self rounded printOn: aStream]
		ifFalse:
			[scaling := 10 raisedToInteger: placesDesired.
			integerPart := numerator abs quo: denominator.
			roundedFractionPart := (numerator abs - (integerPart * denominator)) * scaling * 2 + denominator quo: denominator * 2.
			roundedFractionPart = scaling
				ifTrue:
					[integerPart := integerPart + 1.
					roundedFractionPart := 0].
			"Don't print minus sign if result is rouded to zero"
			(numerator negative and: [integerPart > 0 or: [roundedFractionPart > 0]]) ifTrue: [aStream nextPut: $-].
			integerPart printOn: aStream.
			aStream nextPut: $..
			roundedFractionPart printOn: aStream base: 10 nDigits: placesDesired]!
 !

!Integer methodsFor!

printOn: aStream showingDecimalPlaces: placesDesired
	"Same as super, but provides a faster implementation because fraction part and rounding are trivial."

	self printOn: aStream base: 10.
	placesDesired <= 0 ifTrue: [^self].
	aStream nextPut: $..
	placesDesired timesRepeat: [aStream nextPut: $0]!
 !

!Float methodsFor!

printOn: aStream decimalPlaces: anInteger
	"Append the printed representation of the receiver to the <puttableStream>, aStream, rounded to the <integer> number of decimal places, anInteger."

	self asTrueFraction printOn: aStream showingDecimalPlaces: anInteger!
 !
```

I didn't test it extensively so I'm not confident about this implementation but seems to work ok. Also I'm not recommending anything here, maybe there are better ways to do this. One of the things that would be really useful would be the ability to print numbers with thousands separator, I made my own version but it's slow so I need work on it.

**Comment by blairmcg** [comment]
Unfortunately performing an upfront conversion using `asTrueFraction` is going to give you too much precision in some cases, e.g. 1.0e306. This will print correctly with the original implementation:

```smalltalk
       String streamContents: [:stream | 1.0e306 printOn: stream decimalPlaces: 3] '1000000000000000017216064596736454828831087825013238982328892017892380671244575047987920451875459594568606138861698291060311049225532948520696938805711440650122628514669428460356992624968028329550689224175284346730060716088829214255439694630119794546505512415617982143262670862918816362862119154749127262208.000'
```

But with your version:
```smalltalk
       String streamContents: [:stream | 1.0e306 printOn: stream decimalPlaces: 3] '1000000000000000017216064596736454828831087825013238982328892017892380671244575047987920451875459594568606138861698291060311049225532948520696938805711440650122628514669428460356992624968028329550689224175284346730060716088829214255439694630119794546505512415617982143262670862918816362862119154749127262208.000'
```

There will be other less extreme cases that you might actually care about where there are issues with precision, e.g. 1.25e22. 

```smalltalk
String streamContents: [:s | 1.25e22 printOn: s decimalPlaces: 2]. '12499999999999998951424.00' "Oops"
```

It also won't work as is for the special cases of negative zero, infinities, and nans, so those would need special treatment - it fails for these against the existing `FloatTests`.

Interestingly the 1.0e306 example fails with an error in Pharo 9.0. It won't print `Float fmax` either due to the same issue. Pharo will print 1.0e304 to 2 decimal places, but with too much precision so the result is (arguably) incorrect. Both print correctly in Squeak. The Squeak FP code has diverged a bit from Pharo (or vice versa), but Squeak seems more correct in the detail.

If the base implementation were optimised, we'd have to be careful to make sure it works correctly for all cases to avoid regressions. Correctness with extremes is often less important in applications where the extremes are often not encountered or not interesting, so you might be best served by just creating an optimised version that you have tested for the ranges you care about. If you are happy with the Pharo results, you could just add it in as extension methods, especially as the selector is conveniently different already.

As for printing with thousands separators, you can use `Locale` to do that for you:

```smalltalk
String streamContents: [:strm | Locale default displayFloat: 123456789.0123 on: strm].	"'123,456,789.01'"
```

Results vary depending on the number formatting in place in the locale.

```smalltalk
String streamContents: [:stream | (Locale lcid: 11274) displayFloat: 123456789.0123 on: stream]    "123.456.789,01"
```

If you want a specific format then `printFloat:on:format:` can be used. See the [NUMBERFMT documentation](https://docs.microsoft.com/en-us/windows/win32/api/winnls/ns-winnls-numberfmtw) for more info.

I don't know how acceptable the performance would be for you, however. This has to first print the Float into a canonical string representation, and then call the WinNls GetNumberFormat API (twice). It should still be faster than `printOn:decimalPlaces:` currently is, however, because it doesn't use the slow exact printing logic, but rather rounds to a maximum of 16 significant figures.

**Comment by fxgallego** [comment]
Thanks Blair.
I didn't know about Locale printing so well... I think the performance is "good enough" for my use case.
I was working using `Float>>printOn:decimalPlaces:` and it's noticeable slower than printing using the windows Locale, but I don't care.
The question is why `Float>>printOn:decimalPlaces:` does not use the same logic in Locale current? It is more accurate?

**Comment by fxgallego** [comment]
There is an issue related to Float printing, but not performance. `Float>>printOn:base:significantFigures:` uses a really small range for `decimalExponents: ##(-3 to: 5)`. We had to patch it to ##(-4 to: 14) but I hate to have it patched. Squeak starts at -4 and seems to have no limit.... ¿Could be this setting configurable so we don't have to patch?

**Comment by blairmcg** [comment]
> Thanks Blair.
> I didn't know about Locale printing so well... I think the performance is "good enough" for my use case.
> I was working using `Float>>printOn:decimalPlaces:` and it's noticeable slower than printing using the windows Locale, but I don't care.
> The question is why `Float>>printOn:decimalPlaces:` does not use the same logic in Locale current? It is more accurate?

That's a good question. Printing of floats like this was added to Locale in a2d0d522231454fd0c0c39272d152a730e9633b3 about a year ago as an incidental when improving the support for printing dates and times in locale sensitive formats. There are good tests for the date/time printing, but I omitted to add any tests for the Float formatting. Partly because it is a wrapper around a windows API, and partly because I was either too lazy or forgot to add them. There may be extreme cases that the Locale version doesn't handle properly, and it is in need of some test cases to try that out. 

I think more to the point is that it is a flexible formatting API that is much more awkward to call for printing in standard Smalltalk literal format to variable numbers of decimal places. The overhead of instantiating the NUMVERFMT structure would be relatively high, although I've not measured it relative to the alternative. It's worth considering consolidating all this.

**Comment by blairmcg** [comment]
> There is an issue related to Float printing, but not performance. `Float>>printOn:base:significantFigures:` uses a really small range for `decimalExponents: ##(-3 to: 5)`. We had to patch it to ##(-4 to: 14) but I hate to have it patched. Squeak starts at -4 and seems to have no limit.... ¿

What are the equivalents in Squeak and Pharo of that method? Much of the underlying code is the same (and I believe the interval is exactly the same in Squeak, although expressed slightly differently), but I don't see an apparent public API to print to a number of significant figures. I seem to get the same results in Pharo, Squeak and Dolphin for the basic print operation though:

```smalltalk
#(1.2345e-5 1.2345e-4 1.2345e5 1.2345e6) collect: [:each | each printString] "#('1.2345e-5' '0.00012345' '123450.0' '1.2345e6')"
```

> Could be this setting configurable so we don't have to patch?

Certainly putting the interval in a class variable seems like a worthwhile change.

**Comment by fxgallego** [comment]
>What are the equivalents in Squeak and Pharo of that method? Much of the underlying code is the same (and I believe the interval is exactly the same in Squeak, although expressed slightly differently), but I don't see an apparent public API to print to a number of significant figures. I seem to get the same results in Pharo, Squeak and Dolphin for the basic print operation though:

Sorry, I didn't pay attention and I was evaluating `1e6`. In JavaScript the range is -6..20
It is not enough to change only `Float>>printOn:base:significantFigures:`, there is other range test in `Float>>absPrintExactlyOn:base:`. That forced us to patch also `Float>>printOn:` to use `printOn:base:significantFigures:`. 

> I think more to the point is that it is a flexible formatting API that is much more awkward to call for printing in standard Smalltalk literal format to variable numbers of decimal places. The overhead of instantiating the NUMVERFMT structure would be relatively high, although I've not measured it relative to the alternative. It's worth considering consolidating all this.

Yes, I was thinking in adding a FloatPrinter that takes care of caching the formats for each decimal place. I didn't want to change Locale. But other use cases are with hardcoded decimalSeaparator and thousandsSeparator so the cache should be done by {decimalPlaces. decimalSeparator. thousandSeparator}.

**Comment by blairmcg** [comment]
Above when I said the Locale Float printing code was not well tested, I was wrong. See `WindowsLocaleTest>>testDisplayFloatOn`.

I'm not going to change the `Float>>printOn:` to go through `Float>>printOn:base:significantFigures:` because that must produce an exact (not rounded) result so that literals round trip correctly. We can make it use the same interval setting though.

**Comment by blairmcg** [comment]
One other thing to be aware of is that if you cache a NUMBERFMTW over an image save, then it will be invalid on image restart because the embedded string pointers will be wrong. You may encounter failure (access violiation) of the GetNumberFormat call when experimenting, and this is why. We could probably fix this up if it makes things difficult. Locale objects should also be considered invalid after restart - that's not just because they may contain invalid data (like the NUMBERFMTW), but also because of source the default Locale may have changed, or had user format changes made.

**Comment by fxgallego** [comment]
I'm a bit confused by the comment in `Float class>>defaultDecimalExponents:`, because, it actually uses the Interval provided.

```smalltalk
Float defaultDecimalExponents: (-7 to: 10).
1.2345e-8 printString "'0.000000012345'"

Float defaultDecimalExponents: (-3 to: 5).
1.2345e-8 printString "'1.2345e-8'"
```

**Comment by blairmcg** [comment]
The comment got out of date very fast. I thought I wouldn’t be able to change the printOn: safely, but then found I could, but forgot to go back and fix the comment.

**Comment by fxgallego** [comment]
For `Float>>displayOn:` we have a patch that only prints the float and replaces the '.' by `Locale default decimalSeparator`.

Looking at `NUMBERFMTW>>decimalPlaces:` I see that only works for `0..9` so that is a limiting factor and not really suitable for use with `printOn:decimalPlaces:`

**Comment by blairmcg** [comment]
> For `Float>>displayOn:` we have a patch that only prints the float and replaces the '.' by `Locale default decimalSeparator`.

I have wanted to change `Float>>displayOn:` to go through Locale for a while, and might make it so in 7.2 although I doubt it because it requires quite a lot of work.  The biggest issue with doing so is converting back from the strings so produced in `TypeConverter`s. Windows offers an API to format numbers for display in locale sensitive formats, but not an API to parse them, and the level of variability means that doing so isn't entirely straightforward. For the general case there is rather more to it than just the decimal separator, although that will work well enough for a lot of locales. It would certainly be a breaking change to not have Float's round trip through type converters in existing UIs, and it may be quite impactful anyway since the format would change quite a lot.

> Looking at `NUMBERFMTW>>decimalPlaces:` I see that only works for `0..9` so that is a limiting factor and not really suitable for use with `printOn:decimalPlaces:`, talking about the base system.

Right. I could vaguely remember there were some extra limitations, but not what they were. For most purposes 9 decimal places is probably more than adequate, but yes the base system should only be limited by the precision of the representation.

Just to wrap up, I am not intending to do anything about the performance of `printOn:decimalPlaces:`. There are other, and perhaps better, alternatives if you need something faster. To enumerate some of them for completeness:
- Add a specific C-runtime library API, e.g. `_gcvt` or `_fcvt`, and call that. These are not hard to call and the documentation is good, but format control is limited. Also from my experiments _gcvt seems pretty slow itself.
- Use sprintf as you suggested, which allows quite a lot of control over format.
- Use the `Locale` float printing methods, which as well as being faster adds useful extras such as grouping, e.g. into thousands.
- Define a new set of methods that are based on (for example) inexact float printing, and substantially faster as a result.

**Comment by fxgallego** [comment]
@blairmcg this seems to be a bug. What do you think?

```smalltalk
((283 / 284 * 100.0) roundTo: 0.1) printString -> '99.60000000000001'
```

**Comment by blairmcg** [comment]
No, it's not a bug. 

Multiples of 0.1 cannot be represented precisely as binary fractions, so what you get when you round to 0.1 is a value close to the actual value you may have wanted. If printed with enough precision then the approximation will appear. Float>>printOn:, is designed to print the actual precise value of the Float's bits interpreted as an IEEE FP bit pattern, rather than the value you may have wanted it to be. This is specifically tested for in these cases in `FloatTest>>testPrintString`:

```smalltalk
	"Printed result is exact, so includes any representation error in the Float"
	self assert: (0.1 + 0.2) printString equals: '0.30000000000000004'.
	self assert: (-0.1 - 0.2) printString equals: '-0.30000000000000004'.
	self assert: 240 degreesToRadians cos equals: -0.5000000000000004.
	self assert: 240 degreesToRadians cos abs equals: 0.5000000000000004.
```

I believe you'll get the same result in Pharo and Squeak. 

As I mentioned, Floats are designed to print precisely because they are "literals", so the `printString` must evaluate back to a precisely equal value when compiled, and that means using all the bits of the Float. Ironically in many application situations the fact that Floats cannot exactly represent many common decimal fractions means that you may want to print them with _less_ precision, for example:

```smalltalk
    stream := String writeStream.
    ((283 / 284 * 100.0) roundTo: 0.1) printOn: stream significantFigures: 15.
    stream contents "->" '99.6'
```

**Comment by fxgallego** [comment]
Thanks. I understand the Float logic but I was confused because this changed in master in the last months.
I don't want to bother you with this, I only want to be sure everything is ok compared with previous behavior 😄 , I need to be used to look at the tests before ask. Thanks again!

**Comment by blairmcg** [comment]
It hasn't been changed recently. The current Float printing behaviour using an exact representation (of the bits) dates mostly from the fix for #169, in commit 453537c1d181516c0bf3666112e250aa49056e79 of March 2016, so its been like that for a few years now.
`printOn:significantFigures:`, which is not used for `printOn:`, stopped using the C-runtime library _gcvt function in commit fea581d8bc215541f385f93681db9bfcd0499267 of August 2017, among a bunch of other FP changes.

**Comment by fxgallego** [comment]
Yes you're right again. I'm working with 6 different branches, each one with it's own VM version and with lot of patches on Number 😞. In addition checking against Dolphin master. I'm going **crazy**. I'm removing the patches and writing tests ensuring at least `Float>>printString` behaves the way it was intended in the base Dolphin.
